server/src/main/java/org/apache/iotdb/db/engine/snapshot/SnapshotTaker.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.engine.snapshot;
+
+import org.apache.iotdb.commons.conf.IoTDBConstant;
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+import org.apache.iotdb.db.engine.compaction.log.CompactionLogger;
+import org.apache.iotdb.db.engine.modification.ModificationFile;
+import org.apache.iotdb.db.engine.snapshot.exception.DirectoryNotLegalException;
+import org.apache.iotdb.db.engine.storagegroup.DataRegion;
+import org.apache.iotdb.db.engine.storagegroup.TsFileResource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * SnapshotTaker takes data snapshot for a DataRegion in one time. It does so by creating hard link
+ * for files or copying them. SnapshotTaker supports two different ways of snapshot: Full Snapshot
+ * and Incremental Snapshot. The former takes a snapshot for all files in an empty directory, and
+ * the latter takes a snapshot based on the snapshot that took before.
+ */
+public class SnapshotTaker {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnapshotTaker.class);
+  private final DataRegion dataRegion;
+  public static String SNAPSHOT_FILE_INFO_SEP_STR = "_";
+
+  public SnapshotTaker(DataRegion dataRegion) {
+    this.dataRegion = dataRegion;
+  }
+
+  public boolean takeFullSnapshot(String snapshotDirPath)
+      throws DirectoryNotLegalException, IOException {
+    File snapshotDir = new File(snapshotDirPath);
+    if (snapshotDir.exists()
+        && snapshotDir.listFiles() != null
+        && snapshotDir.listFiles().length > 0) {
+      // the directory should be empty or not exists
+      throw new DirectoryNotLegalException(
+          String.format("%s already exists and is not empty", snapshotDirPath));
+    }
+
+    if (!snapshotDir.exists() && !snapshotDir.mkdirs()) {
+      throw new IOException(String.format("Failed to create directory %s", snapshotDir));
+    }
+
+      dataRegion.syncCloseAllWorkingTsFileProcessors();
```

Comment:
When we calls `syncCloseAllWorkingTsFileProcessors`, all data in memtable will be flushed into disk and the tsfile will be sealed. If we don't flush before we taking a snapshot, though we will create a hard link for the unsealed tsfile, the resource file of it does not exist.
In case of we don't need the snapshot of the resource of the working tsfile, I add a parameter `flushBeforeSnapshot` for `takeFullSnapshot`. If it is false, we will not flush the memtable before we take the snapshot.

Reply:
ok